src/utils.js

```diff
@@ -257,3 +267,265 @@ export function getLinkWithFilename(url) {
+    }
+
+    // Fallback to our list of generic selectors and stop in the first we found.
+    let element;
+    for (const selector of this.DEFAULT_ROOT_SELECTOR.fallback) {
+      element = document.querySelector(selector);
+      if (element) {
+        return selector;
+      }
+    }
+
+    console.debug("We were not able to find the root selector.");
+    return null;
+  }
+
+  /**
+   * Return the documentation tool auto-detected.
+   *
+   * Check for all the known documentation tools and return the name of it if found.
+   * Otherwise, it returns `null`.
+   */
+  getDocumentationTool() {
+    if (this.isSphinx()) {
+      return SPHINX;
+    }
+
+    if (this.isMaterialMkDocs()) {
+      return MKDOCS_MATERIAL;
+    }
+
+    if (this.isPelican()) {
+      return PELICAN;
+    }
+
+    if (this.isDocusaurus()) {
+      return DOCUSAURUS;
+    }
+
+    if (this.isAsciiDoctor()) {
+      return ASCIIDOCTOR;
+    }
+
+    if (this.isJekyll()) {
+      return JEKYLL;
+    }
+
+    if (this.isMkDocs()) {
+      return MKDOCS;
+    }
+
+    console.debug("We were not able to detect the documentation tool.");
+    return null;
+  }
+
+  isSphinx() {
+    return (
+      this.isSphinxAlabasterLikeTheme() ||
+      this.isSphinxReadTheDocsLikeTheme() ||
+      this.isSphinxFuroLikeTheme() ||
+      this.isSphinxBookThemeLikeTheme()
+    );
+  }
+
+  isMaterialMkDocs() {
+    return this.isMaterialMkDocsTheme();
+  }
+
+  isDocusaurus() {
+    return this.isDocusaurusTheme();
+  }
+
+  isPelican() {
+    if (
+      document.querySelectorAll('meta[name="generator"][content="Pelican"]')
+        .length === 1
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  isAsciiDoctor() {
+    if (
+      document.querySelectorAll(
+        'meta[name="generator"][content*="Asciidoctor"]',
+      ).length === 1
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  isJekyll() {
+    if (
+      document.querySelectorAll('meta[name="generator"][content*="Jekyll"]')
+        .length === 1
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  isMkDocs() {
+    // MkDocs adds a comment at the end of the file:
+    //     <!--
+    //     MkDocs version : 1.4.2
+    //     Build Date UTC : 2023-07-11 16:08:07.379780+00:00
+    //    -->
+    if (document.lastChild.textContent.includes("MkDocs version :")) {
+      return true;
+    }
+    return false;
+  }
+
+  isSphinxAlabasterLikeTheme() {
+    const selectors = [
+      'link[href^="_static/alabaster.css"]',
+      'link[href^="_static/flask.css"]',
+      'link[href^="_static/jinja.css"]',
+      'link[href^="_static/click.css"]',
+      'link[href^="_static/celery.css"]',
+      'link[href^="_static/babel.css"]',
+      'link[href^="_static/platter.css"]',
+      'link[href^="_static/werkzeug.css"]',
+    ];
+    for (const selector of selectors) {
+      if (document.querySelectorAll(selector).length === 1) {
+        return true;
+      }
+    }
```

Comment:
This looks like a place where we don't need to iterate. The loop is just looking for the first hit in the list of selectors, which querySelectorAll will do for us.

```suggestion
    if (document.querySelectorAll(selectors.join(", ")).length) {
      return true;
    }
```